### PR TITLE
feat(desktop): editable attachments + data-loss fix on message edit

### PR DESCRIPTION
**Pull request by wesbillman**
## Summary

Fixes a data-loss bug AND lands the proper Slack-style attachment-editable edit flow for desktop. Editing a message with an image/video used to delete the attachment from the saved message; this PR makes attachments first-class state in the edit composer (add / remove / replace, alongside text), and fixes follow-up rendering, focus, and tag-shape regressions caught during testing/review.

**Reported by**: Will Pfleger / kalvin in #sprout-bugs (thread root `6a14995049990b2380527d0bbb90c86d08421f4c7c51efb0eefc01c709d6b0b8`).

### 1. Initial data-loss fix (commits 1–3)

Round-trip imeta media through the composer body so existing attachments survive an edit:

- Edit-load: strip trailing `image|video` lines but remember the imeta-derived URLs.
- Edit-submit: re-append imeta lines to the new body before sending.
- Empty body + non-empty media is allowed (caption-less edit). Empty body + empty media is rejected (no edit-as-deletion).
- Helper `formatImetaMediaLine({url, m})` shared between send and edit paths (mime→alt lives in one place).

### 2. Slack-style edit flow (commits 4–5)

Attachments are now first-class composer state on edit, not just preserved metadata:

- Edit-load seeds `pendingImeta` from `editTarget.imetaMedia`. The existing attachment row renders them as removable thumbnails. Paperclip / drop / paste already work in edit mode — no new UI.
- Edit-submit emits a fresh full imeta tag set (kind:40003 carries arbitrary tags). Body still carries the `` markdown lines so the renderer (`markdown.tsx`) draws them — `imetaByUrl` is metadata enhancement, not a node synthesizer.
- `ImetaMedia` widened to full `BlobDescriptor` (sha256/size/dim/blurhash/thumb/duration preserved).
- Receiver overlay (`applyEditTagOverlay.mjs`): on render, edit's imeta tags replace original's; non-imeta tags (`h`, mentions, replies) are preserved.
- Edit-mode entry preserves any draft attachments the user already had pending in the composer (regression caught in review).
- Rust validation in `edit_message`: empty body + zero attachments rejected (was: empty body alone).

### 3. Refactor pass (commit 6)

Behavior-preserving consolidation:

- `applyEditTagOverlay.mjs` (+ `.d.mts`) — single source for tag overlay, imported by both the renderer and the cache update + tests.
- `buildOutgoingMessage(body, pendingImeta)` — shared by send and edit submit paths (`undefined` = "omit imeta tags" for sends, `[]` = "wipe attachments" for edits).
- `preEditSnapshotRef: { content, pendingImeta }` collapses two refs that were always set/cleared together.

### 4. Focus fix on edit entry (commits 7–8)

Cosmetic regression caught during local testing — pressing Delete after clicking Edit was escaping the composer:

- Root cause: Radix `<DropdownMenuContent>` schedules `setTimeout(0)` to restore focus to the menu trigger on close, which fires AFTER `requestAnimationFrame`. Our edit-mode focus call landed (briefly visible flicker) and was then stolen back to the trigger button.
- Fix: per-mount `editJustSelectedRef` flipped on the Edit `onClick`. `onCloseAutoFocus` reads it and `preventDefault()`s only when Edit was the picked item — Reply / Mark unread / Delete / Copy / reactions / Escape / outside-click all keep their default Radix focus restoration (a11y intact).
- Belt-and-suspenders: the composer's RAF stays in place for any future trigger UI that doesn't override.

### 5. Sparse-imeta hygiene (commit 9)

Reviewer (Codex bot) caught an asymmetry: `buildImetaTags` was emitting `x` and `size` lines unconditionally, so a legacy/cross-client imeta missing those NIP-94 fields would surface as literal `x ` (empty value) and `size 0` on the kind:40003 edit — both rejected by Sprout's relay validator.

- Fix: emit `x` only when `sha256` is non-empty; emit `size` only when `> 0`. Other optional fields (`dim`/`blurhash`/`thumb`/`duration`/`image`) were already conditional.
- Send path is unchanged — full BlobDescriptor always carries valid sha256 + positive size, both lines emitted as before.
- Scope: this is desktop-side structural correctness only. Successfully editing a legacy/cross-client sparse-imeta message ALSO requires relaxing the relay validator (`crates/sprout-relay/src/handlers/imeta.rs` requires `url + m + x + size` AND a local `/media/...` URL). That's a separate follow-up tied to federation / cross-client interop.

## Steps to reproduce the bug (before this PR)

1. Send a message with an image or video attachment.
2. Right-click → Edit on your own message.
3. Change the caption (or save unchanged).
4. **Bug A** (original): the attachment disappears from the saved message.
5. **Bug B** (caught during local testing of A's fix): even with imeta tags preserved, the rendered message shows no media until you re-open the editor.
6. **Bug C** (caught while testing B's fix): clicking Edit briefly focuses the composer, then loses focus to the message row — hitting Delete escapes instead of editing.

## Steps to verify the fix

**P0 — bugs we fixed**
1. Send a message with an image. Edit, change the text, save. → Image still renders. Reload. → Still renders.
2. Click Edit on a message. → Composer holds focus (no flicker, no escape on Delete keypress).
3. Stage a draft image in the composer. Click Edit on an older message. Cancel. → Draft image survives.

**P1 — new capability**
4. Edit a text-only message, paste an image, save. → Image renders.
5. Edit a message with an image, click X on the thumbnail in the composer, save. → Image is gone. Reload. → Still gone.
6. Edit, remove image A, add image B, save. → Only B renders.
7. Send a message with 2 images, edit, remove one, save. → Only the kept one renders.
8. Edit, clear text, keep image, save. → Allowed; renders just the image.
9. Edit, clear text AND remove all attachments. → Save is rejected (no edit-as-deletion).

**P2 — interactions**
10. Repeat #1, #4, #5 with a video.
11. Reactions on a message stay after edit.
12. Mentions / `p` / `h` tags survive the edit.

## Automated test coverage

- `desktop/src/features/messages/lib/imetaMediaMarkdown.test.mjs` — pure helpers (extract, strip, append, round-trip, edge cases) + `buildOutgoingMessage` cases + sparse-imeta projection / round-trip / conditional emit.
- `desktop/src/features/messages/lib/applyEditTagOverlay.test.mjs` — tag overlay (no-edit pass-through, A,B→A,C swap with non-imeta preserved, edit-strips-all, edit-adds-to-text-only, leak-prevention).

**Test counts at HEAD**: Desktop `node:test` 279 / 0 · Rust `src-tauri --lib` 374 / 0. `pnpm typecheck` ✅, `pnpm check` (biome + file-sizes) ✅, full pre-push hooks ✅.

## Out-of-scope follow-ups

- **Mobile parity** — mobile's edit path doesn't handle attachments today.
- **Relay-side imeta validator** — to actually accept legacy/cross-client sparse-imeta edits and federated (non-local URL) media, the validator in `crates/sprout-relay/src/handlers/imeta.rs` needs to allow optional fields and non-local URLs. Tracked separately.
- **NIP-92 `alt` / `fallback` / `service` projection** — currently dropped because `BlobDescriptor` doesn't carry them. Documented at `imetaMediaFromTags`.
- **`MessageComposer.tsx`** is now ~800 lines — flagged as a future split candidate.

## Commits

1. `fix(desktop): preserve attachments when editing messages` — initial round-trip.
2. `fix(desktop): cover edit-attachment edge cases` — caption-less guard, append dedup, legacy-mime fallback.
3. `refactor(desktop): share imeta line formatter between send and edit paths`.
4. `feat(desktop): Slack-style attachment-editable edit flow`.
5. `fix(desktop): address Beth review on attachment-editable edit flow` — render bug + draft-loss.
6. `refactor(desktop): tighten edit-attachment helpers` — overlay extraction + outgoing-message helper + ref collapse.
7. `fix(desktop): focus editor on edit-mode entry so global keybinds land`.
8. `fix(desktop): suppress Radix focus-restoration when picking Edit` — scoped per-Edit-item, a11y preserved on other paths.
9. `fix(desktop): omit imeta x/size lines when values absent` — sparse-imeta hygiene.

### CODE REVIEW

**Comment by wesbillman** [comment]
@codex please review

File: desktop/src/features/messages/lib/imetaMediaMarkdown.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  if (!tags || tags.length === 0) return [];
+  const entries = parseImetaTags(tags as string[][]);
+  const out: ImetaMedia[] = [];
+  for (const entry of entries.values()) {
+    if (!entry.url) continue;
+    out.push({
+      url: entry.url,
+      type: entry.m ?? "image/jpeg",
+      sha256: entry.x ?? "",
+      size: entry.size ?? 0,
```
[the remark this answers is not in the document]
Comment (wesbillman) [reply desktop/src/features/messages/lib/imetaMediaMarkdown.ts:61]: Good catch — addressed in `8a693907`.

`buildImetaTags` now emits `x` only when `sha256` is non-empty, and `size` only when it's a positive number. `url` and `m` stay unconditional (`m` has an `image/jpeg` fallback in `imetaMediaFromTags`, BlobDescriptor's `type` is always populated by the upload pipeline).

Scope clarification: this commit fixes the desktop-side structural emit. Successfully editing a legacy/cross-client sparse-imeta message also requires loosening the relay validator (`crates/sprout-relay/src/handlers/imeta.rs` requires `url + m + x + size` AND a local `/media/...` URL). That's tracked as a separate follow-up — a federation/cross-client edit needs the relay-side change to actually land. With this commit alone, the desktop client stops fabricating `x ` / `size 0` lines; failure mode shifts from "x must be 64-char hex" to a cleaner "missing required field" rejection.

Tests: 5 new cases in `imetaMediaMarkdown.test.mjs` covering sparse projection (`imetaMediaFromTags` with no `x` / no `size`), conditional emit (`buildImetaTags` omits `x` when sha256 empty, omits `size` when 0), and round-trip determinism (sparse imeta → projection → rebuild yields same tag).
